dokka-integration-tests/gradle/build.gradle.kts

```diff
@@ -1,43 +1,208 @@
+    systemProperty(
+        "junit.jupiter.tempdir.cleanup.mode.default",
+        dokkaBuild.isCI.map { isCi -> if (isCi) "ALWAYS" else "ON_SUCCESS" }.get(),
+    )
 
     environment("DOKKA_VERSION", project.version)
+    environment("isExhaustive", dokkaBuild.integrationTestExhaustive)
+    environment("ANDROID_HOME", dokkaBuild.androidSdkDir.get().invariantSeparatorsPath)
+
+    testLogging {
+        exceptionFormat = FULL
+        events(SKIPPED, FAILED)
+        showExceptions = true
+        showCauses = true
+        showStackTraces = true
+    }
+
+    // TODO remove this in https://github.com/Kotlin/dokka/pull/3433
+    doNotTrackState("uses artifacts from Maven Local")
+}
+
+@Suppress("UnstableApiUsage")
+testing {
+    suites {
+        withType<JvmTestSuite>().configureEach {
+            useJUnitJupiter()
+
+            dependencies {
+                implementation(project())
+            }
 
-    inputs.dir(file("projects"))
+            targets.configureEach {
+                testTask.configure {
+                    doFirst {
+                        logger.info("running $path with javaLauncher:${javaLauncher.orNull?.metadata?.javaRuntimeVersion}")
+                    }
+                }
+            }
+        }
+
+        /**
+         * Create a new [JvmTestSuite] for a Gradle project.
+         *
+         * @param[projectPath] path to the Gradle project that will be tested by this suite, relative to [templateProjectsDir].
+         * The directory will be passed as a system property, `templateProjectDir`.
+         */
+        fun registerTestProjectSuite(
+            name: String,
+            projectPath: String,
+            jvm: JavaLanguageVersion? = null,
+            configure: JvmTestSuite.() -> Unit = {},
+        ) {
+            val templateProjectDir = templateProjectsDir.dir(projectPath)
+
+            register<JvmTestSuite>(name) {
+                targets.configureEach {
+                    testTask.configure {
+                        // Register the project dir as a specific input, so changes in other projects don't affect the caching of this test
+                        inputs.dir(templateProjectDir)
+                        // Pass the template dir in as a property, it is accessible in tests.
+                        systemProperty("templateProjectDir", templateProjectDir.asFile.invariantSeparatorsPath)
+
+                        if (jvm != null) {
+                            javaLauncher = javaToolchains.launcherFor { languageVersion = jvm }
+                        }
+                    }
+                }
+                configure()
+            }
+        }
+
+        // register a separate test suite for each 'template' project
+        registerTestProjectSuite(
+            "testTemplateProjectAndroid",
+            "it-android-0",
+            jvm = JavaLanguageVersion.of(11), // AGP requires JVM 11+
+        )
+        registerTestProjectSuite("testTemplateProjectBasic", "it-basic")
```

Comment:
minor/question: as we now have a separate folder for tests with specific template, may be it will be a good idea to move also a template project there?
From one side it could be a little harder to find template project, from the other side it's convenient when test file and project will be near each other.

Or may be at least to align names of template projects and sourceSets to use one convention - up to you

Reply:
Yes, the organisation crossed my mind too!

I see two options

1. keep the template projects as they are now, in a specific project directory
    
    ```
    .
    └── dokka-integration-tests/gradle/
        ├── projects/
        │   ├── it-android-0
        │   ├── it-basic
        │   ├── it-basic-groovy
        │   └── ...
        └── src/
            ├── testTemplateProjectAndroid/
            ├── testTemplateProjectBasic/
            └── testTemplateProjectBasicGroovy/
    ```


3. move the template projects to be in the source
    
    ```
    .
    └── dokka-integration-tests/gradle/
        └── src/
            ├── testTemplateProjectAndroid/
            │   ├── kotlin/
            │   └── project/
            ├── testTemplateProjectBasic/
            │   ├── kotlin/
            │   └── project/
            ├── testTemplateProjectBasicGroovy/
            │   ├── kotlin/
            │   └── project/
            └── ...
    ```


I wasn't sure what would be better, both have pros/cons. I slightly prefer option 1 because it's easier to compare/contrast the template projects, but only slightly. I decided to minimize the changes in this PR.

Reply:
Yeah, let's not change that in this PR, but it could be improved later.
I don't know what is better (separate folder with templates or project in sources of test suite), but I think we can at least align names of test suites (sourceSets) and template projects.
F.e
* templateFolder=`basic` -> sourceSet=`basicTemplateTest` (or `testBasicTemplate` or `testTemplateBasic`)
* templateFolder=`multimodule` -> `sourceSet=`multimoduleTemplateTest` (or ...)
* templateFolder=`multimoduleSpecificCase` -> `sourceSet=`multimoduleSpecificCaseTemplateTest` (or ...)
comparing to current:
* templateFolder=`it-basic` -> sourceSet=`testTemplateProjectBasic`,
* templateFolder=`it-multimodule` -> `sourceSet=`testTemplateProjectMultimodule`
* templateFolder=`it-multimodule-specific-case` -> `sourceSet=`testTemplateProjectMultimoduleSpecificCase`
Though, not in this PR.

Reply:
Yes, agreed to both! Names could be better, and not in this PR.

Reply:
BTW I just had a discussion with @ALikhachev about how to improve the performance - specifically how to reduce the memory usage. One cause of high memory usage is that many of the tests use multiple different versions of Gradle, which means they each have a separate KotlinCompileDaemon.

Something that might help is to change limit what versions of Gradle run concurrently. One idea we had was to 1. move the tests into `src/main`, and 2. generate a test-suite for each Gradle/Kotlin/Android version. That way Gradle will only launch one task per Gradle version.

Reply:
> Something that might help is to change limit what versions of Gradle run concurrently.

Up to you, will wait for the best solution, not sure what is the best approach here with all combination of so much different parameters.
But keep in mind, that some tests are run for different versions of Gradle+Kotlin combination, so looks like it will be a lot of combinations :)
Also, don't forget about IDEA (run/debug) experience if it's possible.

Reply:
> the tests use multiple different versions of Gradle, which means they each have a separate KotlinCompileDaemon.

Not exactly. Each Kotlin version will have its own separate Kotlin daemon, but e.g. in combinations `Gradle 7.0 – Kotlin 1.7.0` and `Gradle 7.5 – Kotlin 1.7.0` there will be single Kotlin daemon and two Gradle daemons